### ISSUE TITLE
JITServer compilation failures when using encryption

### ISSUE THREAD

**Issue by AlexeyKhrabrov** [eclipse-openj9/openj9#15396]
Using JITServer with encryption can result in intermittent remote compilation failures with the error code `compilationStreamFailure` that are caused by SSL I/O errors. The main issue is that such failures are handled by the client as a lost connection, and compilations are performed locally at the client for a short period of time, which leads to memory usage spikes (that are supposed to be eliminated when using JITServer). This happens during normal operation, without any actual network errors.

With the default settings, the client resumes remote compilations 1 second after "losing" the connection, which results in dozens to hundreds of compilations performed locally. If the client is forced to reconnect immediately (using the `-Xjit:reconnectWaitTimeMs=0` option plus the fix in #15391), all compilations are performed remotely - the failed ones are immediately retried remotely and succeed. Increasing the client and server socket timeouts makes no difference.

Here are the SSL errors that I've seen printed to `stderr`, sorted from most to least frequent:
- at the client:
```
error:140943FC:SSL routines:ssl3_read_bytes:sslv3 alert bad record mac:../ssl/record/rec_layer_s3.c:1528:SSL alert number 20
error:140943F2:SSL routines:ssl3_read_bytes:sslv3 alert unexpected message:../ssl/record/rec_layer_s3.c:1528:SSL alert number 10
error:140940F4:SSL routines:ssl3_read_bytes:unexpected message:../ssl/record/rec_layer_s3.c:1461:
```
- at the server:
```
error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:../ssl/record/ssl3_record.c:677:
error:140E0197:SSL routines:SSL_shutdown:shutdown while in init:../ssl/ssl_lib.c:2078:
error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:355:
error:140940F4:SSL routines:ssl3_read_bytes:unexpected message:../ssl/record/rec_layer_s3.c:1461:
```
Not all occurrences of `compilationStreamFailure` at the client have a matching SSL error printed to `stderr`.

The issue is reliably reproducible for me with AcmeAir+JMeter with multiple clients (e.g. 16), even when the server and the clients run on the same machine. I'm using Ubuntu 18.04 with OpenSSL 1.1.1. The issue is reproducible even with a single client compilation thread (although less frequently), thus this is probably not a multithreading issue.

Correctly closing the SSL connections (#15392) seems to get rid of the errors printed to `stderr` and makes the issue less frequent, but doesn't fix it completely. One possibility is that failed `BIO_read()` and `BIO_write()` calls need to be retried in some cases when indicated by `BIO_should_retry()` (similar to how it's done in #15217). Also see #15394.

**Comment by cjjdespres** [comment]
I believe I can reproduce this: I reliably get a few error messages like

```
#JITServer: compThreadID=24 JITServer StreamFailure: JITServer I/O error: read error
#JITServer: compThreadID=24 has failed to compile: compErrCode 48 org/apache/cxf/transport/http/AbstractHTTPDestination.flushHeaders(Lorg/apache/cxf/message/Message;Z)Ljava/io/OutputStream;
```

in the JITServer verbose log under similar conditions (the error code for `compilationStreamFailure` is 48), with corresponding `compilationStreamFailure` messages at the client. Next I'll see if `BIO_should_retry` is `true` when these errors occur.

**Comment by cjjdespres** [comment]
`BIO_should_retry` is in fact true when these errors occur. This is a little puzzling, since the documentation for `BIO_should_retry` states that

> SSL BIOs are the only current exception to this rule: they can request a retry even if the underlying I/O structure is blocking, if a handshake occurs during a call to BIO_read(). An application can retry the failed call immediately or avoid this situation by setting SSL_MODE_AUTO_RETRY on the underlying SSL structure.

The rule they refer to is that BIOs with blocking underlying sockets won't request a retry. I confirmed with some scattered print statements that `SSL_MODE_AUTO_RETRY` is set in both the `SSL_CTX` and the `SSL` in the BIO itself. That means that this shouldn't be happening, since the JITServer BIOs are all blocking.

In any case, I'll see if the problem goes away if we simply continue to try to read from the BIO after this particular type of failed read.

**Comment by cjjdespres** [comment]
Setting `-XX:JITServerTimeout=50000` also seems to fix the issue. (That option controls the socket timeouts in milliseconds for JITServer network communication - the default is 2000). There are still occasional read errors in the JITServer itself like I mentioned in #15505, but they don't appear to affect the clients. It also eliminates the `compilationStreamFailure` errors in the clients due to `EAGAIN` on `read` that I saw when trying to reproduce #15394. This means that the cause of both seem to be timeouts in calls to `read` in the client (directly, or via `BIO_read`) due to high JITServer load.

Possible solutions:

- Increase the default timeout, or disable it entirely. I haven't played around with different values of `JITServerTimeout`, so it may not need to be increased by as much as I did above. Perhaps the socket timeout could be increased dynamically if the client thinks that the server is under too high of a load? I'm not sure if that information is easy to obtain by the client.

- Retry the read (possibly more than once) as I do in #15505, though I'll want to extend that PR to include retrying non-SSL reads too, now that I know the two issues are the same. A cap could also be put on the number of retries that are attempted before failing. I think this would have the same effect as as increasing or disabling the timeout on the socket, however, just in a more roundabout way.
